subprojects/core-api/src/main/java/org/gradle/api/vcs/SourceControl.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.vcs;
+
+import org.gradle.api.Action;
+import org.gradle.api.Incubating;
+
+/**
+ * In settings.gradle:
+ * <pre>
+ * sourceControl {
+ *    repositories {
```

Comment:
I'd go with something more mapping-like, where "mapping" means a function that given a component selector of some kind returns either the VCS URL for that component or "don't know".

In other words, this should look more like the other kinds of component meta-data rules we have, where I can either provide the function (as an `Action` that takes some details) or a hard-coded mapping ("map this component id to this VCS URL").

For example:

```
components {
   withModule("group:name") { 
       // If you need something with maven group and artifact id (group, name), then you can find it here:
       vcsUrl = "https://..."
   }
   all { details ->
      if (details.group == "my-group") {
         vcsUrl = "https://git.my.org/${details.name}"
      }
   }
}
```